zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/clock/ClockPinProcessor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.clock;
+
+import io.camunda.zeebe.engine.processing.distribution.CommandDistributionBehavior;
+import io.camunda.zeebe.engine.processing.streamprocessor.DistributedTypedRecordProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.SideEffectWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedResponseWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.protocol.impl.record.value.clock.ClockRecord;
+import io.camunda.zeebe.protocol.record.intent.ClockIntent;
+import io.camunda.zeebe.stream.api.StreamClock.ControllableStreamClock;
+import io.camunda.zeebe.stream.api.records.TypedRecord;
+import io.camunda.zeebe.stream.api.state.KeyGenerator;
+import java.time.Instant;
+
+public final class ClockPinProcessor implements DistributedTypedRecordProcessor<ClockRecord> {
+  private final SideEffectWriter sideEffectWriter;
+  private final StateWriter stateWriter;
+  private final KeyGenerator keyGenerator;
+  private final ControllableStreamClock clock;
+  private final CommandDistributionBehavior commandDistributionBehavior;
+  private final TypedResponseWriter responseWriter;
+
+  public ClockPinProcessor(
+      final Writers writers,
+      final KeyGenerator keyGenerator,
+      final ControllableStreamClock clock,
+      final CommandDistributionBehavior commandDistributionBehavior) {
+    sideEffectWriter = writers.sideEffect();
+    stateWriter = writers.state();
+    this.keyGenerator = keyGenerator;
+    responseWriter = writers.response();
+    this.clock = clock;
+
+    this.commandDistributionBehavior = commandDistributionBehavior;
+  }
+
+  @Override
+  public void processNewCommand(final TypedRecord<ClockRecord> command) {
+    final long eventKey = keyGenerator.nextKey();
+    final var clockRecord = command.getValue();
+
+    applyClockModification(eventKey, clockRecord);
+    if (command.hasRequestMetadata()) {
+      responseWriter.writeEventOnCommand(eventKey, ClockIntent.PINNED, clockRecord, command);
+    }
```

Comment:
:thought_balloon: This looked a bit odd so I looked into it:
1. It's necessary in case the clock pin command can trigger internally, without a user request. This will most likely not be the case but we could still keep it for future-proofing.
2. The `responseWriter` could do this for us. In fact that's what you [suggested](https://github.com/camunda/camunda/issues/15649#issuecomment-1861566786) when it first came up but [the fix](https://github.com/camunda/camunda/pull/15783) didn't do that, instead adding these manual checks :shrug:

Reply:
Yeah, it seems obvious to me, but :shrug: As I'm not so much in the engine anymore, I'll just follow the rules here :)